### PR TITLE
Handle both prime and binary curves in the SM2 code

### LINKED ISSUE #6646: SM2 code assumes a prime field curve
The SM2 code seems to assume that a prime field curve is in use rather than a binary field:
https://github.com/openssl/openssl/blob/master/crypto/sm2/sm2_crypt.c#L51-L52

Is that a restriction of SM2, or is this incorrect?

Spotted by @dot-asm in this [comment](https://github.com/openssl/openssl/pull/6592#discussion_r199538789).

Ping @randombit, @ronaldtse.

### PR DESCRIPTION
**Pull request by mattcaswell**
Use the correct form of the various *_GFp and *_GF2m functions in the SM2
code based on the type of curve that we have.

Fixes #6646

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by randombit** [comment]
Seems fine to me in so far as I am not aware of any differences in SM2 wrt using a prime vs binary curve. But the machine translation of the standards which I've been going on could easily mangle any such statement. So might be worth checking the test vector in appendix A.3 of http://dacas.cn/sharedimages/ARTICLES/SMAlgorithms/SM2_2.pdf

**Comment by richsalz** [comment]
Thank you for the confirmation! Matt, merge. :)

**Comment by mattcaswell** [comment]
@romen I agree that that is probably the best long term solution. I am reluctant though to go down that route at this stage, i.e. introduce a new API this late in the release cycle. Perhaps we should just open a new issue to be fixed post-1.1.1 that does it that way?

**Comment by romen** [comment]
@mattcaswell sounds good!

(Although I don't really think this qualifies as a "new" API: those functions are already doing exactly the same thing, only throwing equivalent errors with different names! So I could see this change as a fix to reduce code duplication. Especially considering that the current pattern has no benefit, and is actually forcing implementations to use wrappers around them to unify them!)

**Comment by mattcaswell** [comment]
> Although I don't really think this qualifies as a "new" API: those functions are already doing exactly the same thing, only throwing equivalent errors with different names! So I could see this change as a fix to reduce code duplication. Especially considering that the current pattern has no benefit, and is actually forcing implementations to use wrappers around them to unify them!)

I'd like other opinions on this please (ping @openssl/committers). Am I being overly cautious, or should we not do this at this late stage?

File: crypto/sm2/sm2_lcl.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <openssl/ec.h>
+
+static ossl_inline int sm2_get_curve(const EC_GROUP *group, BIGNUM *p,
+                                     BIGNUM *a, BIGNUM *b, BN_CTX *ctx)
+{
+    return EC_METHOD_get_field_type(EC_GROUP_method_of(group))
+               == NID_X9_62_prime_field
+           ? EC_GROUP_get_curve_GFp(group, p, a, b, ctx)
+           : EC_GROUP_get_curve_GF2m(group, p, a, b, ctx);
```
Comment (bernd-edlinger) [inline crypto/sm2/sm2_lcl.h]: This needs  `#ifndef OPENSSL_NO_EC2M`

File: crypto/sm2/sm2_lcl.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+static ossl_inline int sm2_get_affine_coordinates(const EC_GROUP *group,
+                                                  const EC_POINT *p,
+                                                  BIGNUM *x,
+                                                  BIGNUM *y,
+                                                  BN_CTX *ctx)
+{
+    return EC_METHOD_get_field_type(EC_GROUP_method_of(group))
+               == NID_X9_62_prime_field
+           ? EC_POINT_get_affine_coordinates_GFp(group, p, x, y, ctx)
+           : EC_POINT_get_affine_coordinates_GF2m(group, p, x, y, ctx);
```
Comment (bernd-edlinger) [inline crypto/sm2/sm2_lcl.h]: here as well.

File: crypto/sm2/sm2_lcl.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+static ossl_inline int sm2_set_affine_coordinates(const EC_GROUP *group,
+                                                  EC_POINT *p,
+                                                  const BIGNUM *x,
+                                                  const BIGNUM *y,
+                                                  BN_CTX *ctx)
+{
+    return EC_METHOD_get_field_type(EC_GROUP_method_of(group))
+               == NID_X9_62_prime_field
+           ? EC_POINT_set_affine_coordinates_GFp(group, p, x, y, ctx)
+           : EC_POINT_set_affine_coordinates_GF2m(group, p, x, y, ctx);
```
Comment (bernd-edlinger) [inline crypto/sm2/sm2_lcl.h]: and here.

**Comment by kroeckx** [comment]
I understand that it would mean that if we change this later that the current errors will never be returned. But I don't see why it can't return the same error (but in a different function) now. That the function changes is something that happens all the time if we refactor code. So I'm not sure that's a reason to do it now or postpone it.

I think checking which error it returned is almost never a good idea.

**Comment by mattcaswell** [comment]
> I understand that it would mean that if we change this later that the current errors will never be returned. But I don't see why it can't return the same error (but in a different function) now. That the function changes is something that happens all the time if we refactor code. So I'm not sure that's a reason to do it now or postpone it.

It's not really about the change in error code (although that would be a side effect of this change). The point is that we would be introducing a new API and recommending that people use it in preference to the existing API very late in the release cycle. I can't actually think of a problem with the proposed change - but I worry about how late in the release cycle this is in case there are consequences that we haven't thought of.

Technically speaking a new API is a feature, so we're not really supposed to do this anyway - although we have introduced new APIs where it is justified as a bug fix. I'm not really sure we can make that justification here.

**Comment by mattcaswell** [comment]
Fix up commit pushed addressing the issues from @bernd-edlinger.

**Comment by kaduk** [comment]
With a very narrowly scoped new API like this, there is not really going to be a technical problem (unless there's a new bug in the implementation, of course), and there's no point in continuing to have beta releases if we aren't able to make changes if needed, so the "late in the release process" part does not really have much weight in my mind.  The relevant question to me seems to be more of whether this is a bugfix or a feature, and since this is not a major release where we have a good chance to deprecate the old stuff, I think that leans me in favor of waiting to put it in 1.1.1a to prepare for future API breakage.

**Comment by mattcaswell** [comment]
Alternative fix in #6815.

**Comment by mattcaswell** [comment]
Closing in favour of #6815 (which just got merged).

**Pull request closed without merging**